### PR TITLE
Openshift unittests

### PR DESCRIPTION
**Pull request by rtnpro**
Tests and related refactor for openshift provider using Openshift API.

### CODE REVIEW

File: atomicapp/providers/openshift.py [hunk 1]
```diff
@@ -179,15 +266,7 @@ def _process_template(self, template):
...
-        (status_code, return_data) = self._make_request("post", url, data=template)
-        if status_code == 201:
-            logger.info("template proccessed %s", template['metadata']['name'])
-            logger.debug("processed template %s", return_data)
-            return return_data['objects']
-        else:
-            msg = "%s %s" % (status_code, return_data)
-            logger.error(msg)
-            raise ProviderFailedException(msg)
+        return self.oc.process_template(url, template)
```
Comment (kadel) [inline atomicapp/providers/openshift.py]: Is `OpenShiftProvider._process_template` still needed?
I would probably   get rid of if this function (`_process_template`) now. It doesn't do anything useful, everything important is in `OpenshiftClient.process_tempate`.

Comment (rtnpro) [inline atomicapp/providers/openshift.py]: ah! yes

**Comment by dustymabe** [comment]
still WIP or done now?

**Comment by rtnpro** [comment]
@dustymabe 

Yes, currently having 78% coverage. I need to cover a few more methods.

File: atomicapp/providers/openshift.py (hunk 1 above)
Comment (rtnpro) [inline atomicapp/providers/openshift.py]: The reason I did not remove ``_process_template`` is because of the dependency on ``get_url``. Also, I see no reason to move ``get_url`` to ``OpenshiftClient``.

File: atomicapp/providers/openshift.py [hunk 2]
```diff
@@ -32,6 +32,85 @@
 logger = logging.getLogger(__name__)
 
 
+class OpenshiftClient(object):
+
+    def __init__(self, openshift_api, kubernetes_api):
+        self.openshift_api = openshift_api
+        self.kubernetes_api = kubernetes_api
+
+    def get_oapi_resources(self):
```
Comment (dustymabe) [inline atomicapp/providers/openshift.py]: could probably make this function generic:

get_api_resources(self, type) where type = 'kubernetes' or 'openshift'

don't have to change it, just sayin

File: tests/units/providers/test_openshift_provider.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
+# -*- coding: utf-8 -*-
```
Comment (dustymabe) [inline tests/units/providers/test_openshift_provider.py]: This file could use some comments about the strategy of the test.

File: tests/units/providers/test_openshift_provider.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    def test_process_artifact_data_error_resource_not_in_resources(self):
+        artifact_data = {
+            'kind': 'foobar'
+        }
+
+        op = self.get_oc_provider()
+
+        self.assertRaises(
+            ProviderFailedException,
+            op._process_artifact_data, 'foo', artifact_data)
```
Comment (dustymabe) [inline tests/units/providers/test_openshift_provider.py]: should this not be:

`op._process_artifact_data('foo', artifact_data))`

File: atomicapp/providers/openshift.py [hunk 5]
```diff
@@ -294,16 +338,24 @@ def _parse_kubeconf(self, filename):
             user:
                 token: abcdefghijklmnopqrstuvwxyz0123456789ABCDEF
         """
```
Comment (dustymabe) [inline atomicapp/providers/openshift.py]: Should move this example expected file format to the _parse_kubeconf_data function now

File: atomicapp/providers/openshift.py [hunk 6]
```diff
@@ -313,6 +365,9 @@ def _parse_kubeconf(self, filename):
             if co["name"] == current_context:
                 context = co
 
+        if not context:
+            raise ProviderFailedException()
```
Comment (dustymabe) [inline atomicapp/providers/openshift.py]: Don't really understand why this change would made. Please change it back if there was not good reason.

File: atomicapp/providers/openshift.py (hunk 2 above)
Comment (rtnpro) [inline atomicapp/providers/openshift.py]: yes.

File: tests/units/providers/test_openshift_provider.py (hunk 3 above)
Comment (rtnpro) [inline tests/units/providers/test_openshift_provider.py]: :P but agreed!

File: tests/units/providers/test_openshift_provider.py (hunk 4 above)
Comment (rtnpro) [inline tests/units/providers/test_openshift_provider.py]: No, this is how ``self.assertRaises`` is supposed to be called.

``self.assertRaises(ExceptionClass, method, arg1, arg2)``

File: atomicapp/providers/openshift.py (hunk 5 above)
Comment (rtnpro) [inline atomicapp/providers/openshift.py]: Yes

File: atomicapp/providers/openshift.py (hunk 6 above)
Comment (rtnpro) [inline atomicapp/providers/openshift.py]: without this check, the subsequent code used to break when ``context == None``, as they check for ``context[somekey]``.

File: tests/units/providers/test_openshift_provider.py (hunk 4 above)
Comment (dustymabe) [inline tests/units/providers/test_openshift_provider.py]: ok

File: atomicapp/providers/openshift.py (hunk 6 above)
Comment (dustymabe) [inline atomicapp/providers/openshift.py]: got ya. you're right

It would still be nice to print out the Invalid filename like was done on line 327 before?

**Comment by dustymabe** [comment]
LGTM. @kadel can you +1 and we can merge this

**Comment by kadel** [comment]
:+1: LGTM
